### ISSUE TITLE
False positive error: pack expansion contains parameter pack '_UTypes' that has a different length (1 vs. 2) from outer parameter packs

### ISSUE THREAD

**Issue by Romain-Geissler-1A** [llvm/llvm-project#61415]
Hi,

This is an issue similar #58452, actually the snippet I posted in this issue is the same. Using libstc++'s 13 tuple implementation is still not possible when using `-std=c++2b`:

```cpp
#include <tuple>

void g(std::tuple<std::tuple<int, int>>);

void f() 
{
    g(std::forward_as_tuple<std::tuple<int, int>>({0, 0}));
}
```

Godbolt output (https://godbolt.org/z/Meca376M6):
```console
In file included from <source>:1:
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:691:2: error: pack expansion contains parameter pack '_UTypes' that has a different length (1 vs. 2) from outer parameter packs
        using __convertible = __and_<is_convertible<_UTypes, _Types>...>;
        ^~~~~
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:853:27: note: in instantiation of template type alias '__convertible' requested here
          = _TCC<true>::template __convertible<_Args...>::value;
                                 ^
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:948:12: note: in instantiation of static data member 'std::tuple<std::tuple<int, int> &&>::__convertible<int &, int &>' requested here
        explicit(!__convertible<_UElements&...>)
                  ^
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:2000:36: note: while substituting deduced template arguments into function template 'tuple' [with _UElements = <int, int>]
    { return tuple<_Elements&&...>(std::forward<_Elements>(__args)...); }
                                   ^
<source>:7:12: note: in instantiation of function template specialization 'std::forward_as_tuple<std::tuple<int, int>>' requested here
    g(std::forward_as_tuple<std::tuple<int, int>>({0, 0}));
           ^
In file included from <source>:1:
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:691:2: error: pack expansion contains parameter pack '_UTypes' that has a different length (1 vs. 2) from outer parameter packs
        using __convertible = __and_<is_convertible<_UTypes, _Types>...>;
        ^~~~~
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:853:27: note: in instantiation of template type alias '__convertible' requested here
          = _TCC<true>::template __convertible<_Args...>::value;
                                 ^
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:959:12: note: in instantiation of static data member 'std::tuple<std::tuple<int, int> &&>::__convertible<const int, const int>' requested here
        explicit(!__convertible<const _UElements...>)
                  ^
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/13.0.1/../../../../include/c++/13.0.1/tuple:2000:36: note: while substituting deduced template arguments into function template 'tuple' [with _UElements = <int, int>]
    { return tuple<_Elements&&...>(std::forward<_Elements>(__args)...); }
                                   ^
<source>:7:12: note: in instantiation of function template specialization 'std::forward_as_tuple<std::tuple<int, int>>' requested here
    g(std::forward_as_tuple<std::tuple<int, int>>({0, 0}));
           ^
2 errors generated.
Compiler returned: 1
```

**Comment by shafik** [comment]
CC @erichkeane

**Comment by erichkeane** [comment]
I'll need a minimal example here, I got lost on the initial one trying to figure out what is going on inside of Tuple.  Also, would be nice to know if perhaps this is just a libc++ bug instead :)

**Comment by Romain-Geissler-1A** [comment]
Note, I didn't try to reduce my case since @stbergmann posted his own reduced C++ snippet. Apparently it's not a regression in clang, as it seems clang never accepted this code in any version (checking on Compiler Explorer). In practice, given the updates made on libstdc++ <tuple> side, it makes the just released clang 16 error when using libstdc++ tuples with C++23.

**Comment by Romain-Geissler-1A** [comment]
The problem still happens with clang 17 rc1.

**Comment by Romain-Geissler-1A** [comment]
Note: with gcc 13 now being released, this bug is more likely to be triggered in real life code. Anyone using libstdc++ >= 13 and C++ >= 23 and trying to use maps using std::tuple will trigger this bug.

**Comment by vient** [comment]
It triggers in real life code with clang 17.0.1 and libstdc++ 13 indeed.

**Comment by Romain-Geissler-1A** [comment]
Hi,

The original reproducer in the first message of this issue seems to no longer happen with the clang trunk (and is still broken with clang 17, so it's not some hidden clang workaround on libstdc++ side which "fixed" the issue).

However the reduced reproducer from @stbergmann here https://github.com/llvm/llvm-project/issues/61415#issuecomment-1470076987 is still broken right now with clang trunk.

Cheers,
Romain

**Comment by Yanpas** [comment]
Did anyone try clang 18.1?

**Comment by Romain-Geissler-1A** [comment]
No, and Compiler explorer doesn't have clang 18 yet. However it works with clang trunk since some months, so there is no reason it shouldn't work with clang 18 (the reproducer from the first message).

However it seems the reduced case from @stbergmann in https://github.com/llvm/llvm-project/issues/61415#issuecomment-1470076987 is still broken on trunk.

**Comment by shafik** [comment]
I think we should close this issue. 

CC @stbergmann can you confirm if your reduced case is the same as https://github.com/llvm/llvm-project/issues/32252

**Comment by vient** [comment]
Original test case works fine with Clang 18.1.
Both https://github.com/llvm/llvm-project/issues/61415#issuecomment-1470076987 and https://github.com/llvm/llvm-project/issues/32252 still get errors with Clang 18.1 though.

**Comment by cor3ntin** [comment]
@mizvekov
